src/Runner.Common/RunServer.cs

```diff
@@ -67,10 +67,48 @@ private void CheckConnection()
             }
         }
 
-        public Task<AgentJobRequestMessage> GetJobMessageAsync(string id)
+        public Task<AgentJobRequestMessage> GetJobMessageAsync(string id, CancellationToken cancellationToken)
         {
             CheckConnection();
-            return _taskAgentClient.GetJobMessageAsync(id);
+            var jobMessage = RetryRequest<AgentJobRequestMessage>(async () =>
+                                                    {
+                                                        return await _taskAgentClient.GetJobMessageAsync(id, cancellationToken);
+                                                    }, cancellationToken);
+            return jobMessage;
+        }
+
+        private async Task<T> RetryRequest<T>(Func<Task<T>> func,
+                                                        CancellationToken cancellationToken,
+                                                        int maxRetryAttempts = 5
+                                                        )
+        {
+            var currentRetryAttempt = 1;
+            while (true)
```

Comment:
I am not sure that it makes sense to duplicate the same condition twice:
1) `catch (Exception ex) when (retryCount < maxRetryAttemptsCount)`
2) `while (retryCount < maxRetryAttemptsCount)`

Actually, condition in `while` will never return false because the exception in `catch ... when ...` will be thrown earlier and having condition in two places can confuse so I prefer to keep the `catch ... when ...` approach + `while (true)`.
I have looked through some retries implementations in different places and actions-dotnet and it seems that `while (true)` approach is widely used.

Also if we put condition to `while`, C# compiler thinks that code after loop can be reached and force to write some `return` after loop. it makes code less readable because it is useless code lines that will never be reached.